src/libawkward/array/NumpyArray.cpp

```diff
@@ -573,8 +573,28 @@ namespace awkward {
     throw std::invalid_argument("array contains no Records");
   }
 
+  const std::vector<ssize_t> flatten_shape(const std::vector<ssize_t> shape) {
+    if (shape.size() == 1) {
+      return std::vector<ssize_t>();
+    }
+    else {
+      std::vector<ssize_t> out = { shape[0]*shape[1] };
+      out.insert(out.end(), shape.begin() + 2, shape.end());
+      return out;
+    }
+  }
+
+  const std::vector<ssize_t> flatten_strides(const std::vector<ssize_t> strides) {
+    if (strides.size() == 1) {
+      return std::vector<ssize_t>();
+    }
+    else {
+      return std::vector<ssize_t>(strides.begin() + 1, strides.end());
+    }
+  }
+
   const std::shared_ptr<Content> NumpyArray::flatten(int64_t axis) const {
-    throw std::runtime_error("FIXME: not implemented");
+    return std::make_shared<NumpyArray>(identities_, parameters_, ptr_, flatten_shape(shape_), flatten_strides(strides_), byteoffset_, itemsize_, format_);
```

Comment:
This needs to be tested in some extreme cases. We should slice up a 2-dimensional array in weird ways to give it odd shape/strides combinations. These `flatten_shape` and `flatten_strides` functions were made for a different purpose (technical step in `getitem`) and might be making some assumptions that would be unwarranted here.

A NumpyArray with odd shape/strides combinations can be compacted (`NumpyArray::compact`) to copy the array into one with completely contiguous shape/strides.

I should explain what I mean by "odd shape/strides combinations." Here's a perfectly normal NumPy array:

```python
>>> import numpy
>>> array = numpy.arange(2*3*5).reshape(2, 3, 5)
>>> array
array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14]],

       [[15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]]])
```

Its `strides[i] == itemsize*product(shape[i + 1:])`, which is to say that there's no unreachable or overlapping elements (in exact analogy with ListArray, though NumPy arrays can describe their striding with a tuple—they don't need two whole `starts`/`stops` arrays).

```python
>>> array.shape
(2, 3, 5)
>>> array.strides
(120, 40, 8)
>>> array.itemsize
8
>>> array.itemsize*array.shape[1]*array.shape[2], array.itemsize*array.shape[2], array.itemsize
(120, 40, 8)
```

Then if we slice it up in strange ways, skipping elements with `1::2` and not going all the way to the ends of an inner array,

```python
>>> array2 = array[:, 1::2, :3]
>>> array2
array([[[ 5,  6,  7]],

       [[20, 21, 22]]])
```

the shapes and strides no longer fit this pattern:

```python
>>> array2.shape
(2, 1, 3)
>>> array2.strides
(120, 80, 8)
>>> array.itemsize
8
>>> array2.itemsize*array2.shape[1]*array2.shape[2], array2.itemsize*array2.shape[2], array2.itemsize
(24, 24, 8)
```

So what I'm asking is for a test of `NumpyArray::flatten` with `awkward1.layout.NumpyArray(array2)` and maybe also `awkward1.layout.NumpyArray(array2[0])`. If it doesn't work, then putting in a check for `NumpyArray::iscompact` and possibly calling `NumpyArray::compact` will be necessary.

We'd like to avoid compacting the array if we can, but if there isn't an easy way to do it, then the work to figure out a clever way isn't justified.

Reply:
Another thing that should be tested: `array3` with a negative step, so that one of the strides is negative. It's possible that `array2` would pass while `array3` would fail—I just haven't thought through those cases.

Reply:
@jpivarski - yes, you are correct. I'll need to look into it. Current implementation fails:
```python
        array3 = array[:, 1::2, :3]
        print(awkward1.tolist(array3))
        assert awkward1.tolist(array3) == [[[5, 6, 7]], [[20, 21, 22]]]
        assert flatten(awkward1.tolist(array3)) == [[5, 6, 7], [20, 21, 22]]
>       assert awkward1.tolist(array3.flatten()) == [[5, 6, 7], [20, 21, 22]]
E       assert [[5, 6, 7], [15, 16, 17]] == [[5, 6, 7], [20, 21, 22]]
E         At index 1 diff: [15, 16, 17] != [20, 21, 22]
E         Full diff:
E         - [[5, 6, 7], [15, 16, 17]]
E         + [[5, 6, 7], [20, 21, 22]]

```

Reply:
> We'd like to avoid compacting the array if we can, but if there isn't an easy way to do it, then the work to figure out a clever way isn't justified.

The following solution works on the examples provided:
```cpp
if (iscontiguous()) {
      return std::make_shared<NumpyArray>(identities_, parameters_, ptr_, flatten_shape(shape_), flatten_strides(strides_), byteoffset_, itemsize_, format_);
    }
    else {
      return contiguous().flatten(axis);
    }
```